includes/class-stats.php

```diff
@@ -59,22 +60,115 @@ public function migrate() {
 				"CREATE TABLE {$wpdb->wpjm_stats} (
 				`id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
 				`date` date NOT NULL,
-				`post_id` bigint(20) DEFAULT NULL,
+				`post_id` bigint(20) DEFAULT 0,
 				`name` varchar(255) NOT NULL,
 				`group` varchar(255) DEFAULT '',
 				`count` bigint(20) unsigned not null default 1,
 				PRIMARY KEY (`id`),
-				INDEX `idx_wpjm_stats_name_date_group`  (`name`, `date`, `group`)
+				INDEX `idx_wpjm_stats_name_date_group`  (`name`, `date`, `group`),
+				UNIQUE INDEX `idx_wpjm_stats_name_date_group_post_id`  (`name`, `date`, `group`, `post_id`)
 			) {$collate}",
 			]
 		);
 	}
 
+	/**
+	 * Log a stat into the db.
+	 *
+	 * @param string $name         The stat name.
+	 * @param string $group        The group this stat belongs to.
+	 * @param int    $post_id      The post_id this stat belongs to.
+	 * @param int    $increment_by The amount to increment the stat by.
+	 *
+	 * @return bool
+	 */
+	public function log_stat( string $name, string $group = '', int $post_id = 0, int $increment_by = 1 ) {
+		global $wpdb;
+
+		if (
+			strlen( $name ) > 255 ||
+			strlen( $group ) > 255 ||
+			! is_numeric( $post_id ) ||
+			! is_numeric( $increment_by ) ) {
+			return false;
+		}
+
+		$date_today = gmdate( 'Y-m-d' );
+
+		// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery
+		$result = $wpdb->query(
+			$wpdb->prepare(
+				"INSERT INTO {$wpdb->wpjm_stats} " .
+				'(`name`, `date`, `group`, `post_id`, `count` ) ' .
+				'VALUES (%s, %s, %s, %d, %d) ' .
+				'ON DUPLICATE KEY UPDATE `count` = `count` + %d',
+				$name,
+				$date_today,
+				$group,
+				$post_id,
+				$increment_by,
+				$increment_by
+			)
+		);
+
+		if ( false === $result ) {
+			return false;
+		}
+
+		$cache_key = $this->get_cache_key_for_stat( $name, $group, $post_id );
+
+		wp_cache_delete( $cache_key, 'wpjm_stats' );
```

Comment:
I think we are not really better off caching individual rows — or do you imagine this being used to invalidate cache for more complex queries?

Reply:
Yesh! -- this is an example. Like instead of invalidating individual rows, we would invalidate the cached result of a more complex query. In reality, probably we would need to replace this code with an action.